File: java/src/org/ggolawski/security/utils/crypto/HMAC.java
package org.ggolawski.security.utils.crypto;

import java.security.GeneralSecurityException;

import javax.crypto.Mac;
import javax.crypto.SecretKey;

import org.apache.commons.codec.binary.Hex;

public class HMAC {
	private static final String ALGORITHM = "HmacSHA256";
	
	public static void main(String[] args) throws GeneralSecurityException {
		String input = "Very secret text";
		System.out.println("Beginning; text to HMAC: " + input);
		
		// Generate random secret key
		System.out.println("Generating secret key...");
		SecretKey key = Utils.generateKey(ALGORITHM, 128);
		System.out.println("Secret key: " + Hex.encodeHexString(key.getEncoded()));
		
		Mac mac = Mac.getInstance(ALGORITHM);
		
		// Generate HMAC
		System.out.println("Generating HMAC...");
		mac.init(key);
		mac.update(input.getBytes());
		byte[] code = mac.doFinal();
		System.out.println("HMAC: " + Hex.encodeHexString(code));
	}
}

File: php/crypto/HMAC.php
<?php
$input = "Very secret text";
echo "Beginning; text to HMAC: " . $input, PHP_EOL;

// Generate random secret key
echo "Generating secret key...", PHP_EOL;
$key = openssl_random_pseudo_bytes(16);
echo "Secret key: " . bin2hex($key), PHP_EOL;

// Generate HMAC
echo "Generating HMAC...", PHP_EOL;
$mac = hash_hmac("sha256", $input, $key);

echo "HMAC: " . $mac, PHP_EOL;
?>